django/contrib/contenttypes/fields.py

```diff
@@ -500,15 +500,41 @@ def get_prefetch_queryset(self, instances, queryset=None):
                     False,
                     self.prefetch_cache_name)
 
-        def add(self, *objs):
+        def add(self, *objs, **kwargs):
+            bulk = kwargs.pop('bulk', True)
             db = router.db_for_write(self.model, instance=self.instance)
-            with transaction.atomic(using=db, savepoint=False):
+
+            def check_and_update_obj(obj):
+                if not isinstance(obj, self.model):
+                    raise TypeError("'%s' instance expected" % self.model._meta.object_name)
+                setattr(obj, self.content_type_field_name, self.content_type)
+                setattr(obj, self.object_id_field_name, self.pk_val)
+
+            if bulk:
+                ids = []
                 for obj in objs:
-                    if not isinstance(obj, self.model):
-                        raise TypeError("'%s' instance expected" % self.model._meta.object_name)
-                    setattr(obj, self.content_type_field_name, self.content_type)
-                    setattr(obj, self.object_id_field_name, self.pk_val)
-                    obj.save()
+                    check_and_update_obj(obj)
+                    if obj._state.adding or obj._state.db != db:
```

Comment:
Now I understand this check doesn't work on the failing test because the objects appear to be saved but were deleted in the meantime by `bacon.tags.set([])`. For now, I fixed the test by adding `bulk=False` to the subsequent `.set()` test: `bacon.tags.set([fatty, salty], bulk=False, clear=True)`. Is that sufficient? I didn't understand what other checks we'd add to reject unsaved objects.